WIP: #763 - Fixes "Chunk A" group of bugs

sdegutis opened the pull request.

sdegutis committed:
[#763] Adding Cypress test for Scenario 1.

sdegutis committed:
[#763] Resolving Scenario 1.

sdegutis committed:
[#763] Adding Cypress test for Scenario 2 which passes.

sdegutis committed:
[#763] Adding Cypress test for Scenario 3 which passes.

sdegutis committed:
[#763] Adding Cypress test for Scenario 4 which fails.

sdegutis committed:
[#763] Splitting Scenario 4 into two tests to match GH issue.

sdegutis committed:
[#763] Putting scenarios into test groups.

sdegutis committed:
[#763] Commenting out code that currently has no effect on any tests.

sdegutis committed:
[#763] Never mind. Something else must have been wrong.

taoeffect commented on frontend/model/state.js:
```diff
@@ -355,6 +355,9 @@ const getters = {
   groupIncomeDistribution (state, getters) {
     return groupIncomeDistribution({ state, getters, monthstamp: currentMonthstamp() })
   },
+  groupAdjustedIncomeDistribution (state, getters) {
+    return groupIncomeDistribution({ state, getters, monthstamp: currentMonthstamp(), adjusted: true })
+  },
```
Why is this being created instead of using the existing `groupIncomeAdjustedDistribution` which does the same thing?

sdegutis replied:
You're probably right, it's probably the same thing. I looked for this but didn't find it at first. I'll take a closer look at this code again Monday.

But the bigger question is about "ourContributionSummary" which is described in [this comment](https://github.com/okTurtles/group-income-simple/issues/763#issuecomment-704541338). That's the real thing I need feedback here, and the main reason I opened this WIP PR.

taoeffect reviewed (commented):
This PR is currently [missing this part](https://github.com/okTurtles/group-income-simple/issues/763#issuecomment-658225618):

>  Now the only thing left is to not make it a partial payment:

The tests should also verify that too, i.e. in Scenario 1 make sure that it's not displayed as a partial payment when it shouldn't be, and fail if it is.

sdegutis committed:
[#763] Skip Scenario 4.1's test for now in Cypress...

sdegutis committed:
[#763] Extract paymentTotalFromUserToUser so it can be used outside Vue.

sdegutis committed:
[#763] Inlining unnecessary variables.

sdegutis committed:
[#763] Extracting non-Vue version of groupIncomeDistribution().

sdegutis committed:
[#763] Extracting object-remapping logic into helper function.

sdegutis committed:
[#763] Wrote first test for group income distribution logic.

sdegutis committed:
[#763] Wrote basic test for adjusted group income distribution logic.

sdegutis committed:
[#763] Cleaning up whitespace.

sdegutis committed:
[#763] Renaming test and adding new test for group income distribution.

sdegutis committed:
[#763] Simplifying interface to paymentTotalFromUserToUser().

sdegutis committed:
[#763] Rewriting some logic to use functional programming.

sdegutis committed:
[#763] Adding comment to explain relative -> absolute conversion.

sdegutis committed:
[#763] Splitting out group income distribution and adjustment logic.

sdegutis committed:
[#763] Whitespace again.

sdegutis committed:
[#763] Turned Scenarios 4 and 4.1 into unit test cases.

sdegutis committed:
[#763] This whitespace is fine to get rid of, from/to becomes clearer.

sdegutis committed:
[#763] Writing down notes that may help us solve failing test case.

sdegutis committed:
[#763] Added unit test for Scenario 3.

sdegutis committed:
[#763] Added unit test for Scenario 2.

sdegutis committed:
[#763] Added unit test for Scenario 1.

sdegutis committed:
[#763] Recombining group income distributions functions to pass test.

sdegutis committed:
[#763] Fixing linter error that silently caused my code to not run.

sdegutis committed:
[#763] Got Scenario 4.1 passing, but Scenario 3 fails now.

sdegutis committed:
[#763] Added Scenario 3's test case to mincome-proportional tests.

sdegutis committed:
[#763] Fixed Scenario 3's test case in mincome-proportional tests.

sdegutis commented:
> This PR is currently [missing this part](https://github.com/okTurtles/group-income-simple/issues/763#issuecomment-658225618):
> 
> > Now the only thing left is to not make it a partial payment:
> 
> The tests should also verify that too, i.e. in Scenario 1 make sure that it's not displayed as a partial payment when it shouldn't be, and fail if it is.

Will work on that next.

In the meantime, feel free to check the code in this PR that solves the 5 scenarios.

sdegutis committed:
[#763] Deleting code that isn't used anymore.

sdegutis committed:
[#763] Cleaning up tests a bit.

sdegutis committed:
[#763] Changing Scenario 3's test to reflect what "should" happen.

sdegutis committed:
[#763] Removing unused import.

sdegutis committed:
[#763] Adding unit test to help explain/test distribution logic.

sdegutis committed:
[#763] Removing unused import that's causing project to not compile.

sdegutis committed:
[#763] Changing quote style that's preventing code from compiling.

sdegutis committed:
[#763] Removing extra comma that's preventing code from compiling.

sdegutis commented:
The next thing in this PR that needs a set of eyes is [this comment](https://github.com/okTurtles/group-income-simple/issues/763#issuecomment-711084166).

sdegutis committed:
[#763] Changing the "Payments sent" numerator to only sent payments.

sdegutis committed:
[#763] Turning first Cypress scenario into unit test.

sdegutis committed:
[#763] Tentatively updating Cypress test to reflect new findings:

See https://github.com/okTurtles/group-income-simple/issues/763#issuecomment-712166451

sdegutis committed:
[#763] Fixed bug in group income distribution logic.

sdegutis committed:
[#763] Turned next failing Cypress test into a unit test.

sdegutis committed:
[#763] Fixing more Cypress payment specs.

sdegutis committed:
[#763] Turning current broken situation into a unit test.

sdegutis committed:
[#763] Putting each expected test result array's items on one line.

sdegutis committed:
[#763] Fixing a better-understood Scenario 3 by reverting code.

sdegutis committed:
[#763] Fixing possibly flaky tests (only fails on precision):

Note, these might not be flaky long-term, they might only be different
now because of the new algorithm code in a recent commit. Since this
commit message will only ever show up in a squashed commit, I mean
that a commit or two ago, I went back to Greg's algorithm initially
posted in the github comments for this message, which I temporarily
took out and tried a different (incorrect) algorithm with.

sdegutis committed:
[#763] Adding member join-date into group income distribution test cases.

sdegutis committed:
[#763] Fixing and skipping second failing test.

sdegutis committed:
[#763] Passing member joinedDate to function from UI.

sdegutis committed:
[#763] Passing full creation timestamp in payments to algorithm.

sdegutis committed:
[#763] Adding helper function to transform data into events.

sdegutis committed:
[#763] Moving helper function above main functions.

sdegutis committed:
[#763] Splitting haves/needs in helper function for convenience.

sdegutis committed:
[#763] Adding another test case for helper function.

sdegutis committed:
[#763] Making withDate argument explicit.

sdegutis committed:
[#763] Fixing indentation.

sdegutis committed:
[#763] Making sure helper function works without adjustWith.

sdegutis committed:
[#763] Starting to rewrite tests using more semantic data.

sdegutis committed:
[#763] Verifying data->event transformations for existing tests.

sdegutis committed:
[#763] Creating sub-test-suite for tests using new helper function.

sdegutis committed:
[#763] Doesn't seem necessary to give new function the mincome.

sdegutis committed:
[#763] Exclude users who have no excess from algorithm.

sdegutis committed:
[#763] Cleaning up test by removing useless variable.

sdegutis committed:
[#763] Wrote new algorithm.

sdegutis committed:
[#763] Converting old tests for new algorithm.

sdegutis committed:
[#763] Current payments are deducted from todo payments.

sdegutis committed:
[#763] Starting to distribute excess payments back to other payments.

sdegutis committed:
[#763] A few other tests already pass.

sdegutis committed:
[#763] A few more tests pass with altered order/precision.

sdegutis committed:
[#763] Refactoring into multiple functions and trying a fix.

taoeffect commented on frontend/utils/giLodash.js:
```diff
@@ -246,3 +246,7 @@ export function deepEqualJSONType (a: any, b: any): boolean {
+
+export function remapObject (o, f) {
+  return Object.fromEntries(Object.entries(o).map(([k, v]) => [k, f(v)]))
+}
```
Nicely done.

taoeffect replied:
> AFAIK there was a destructive version at the top of the file called mapValues which matches the name of it in lodash, but I needed an immutable one. But I could be wrong on both counts.

Two things:

1. I didn't realize that, thank you for pointing it out. Having looked them both over, it seems that indeed `remapObject` looks to be a re-implementation of `mapValues`, and I'm guessing it's slower because of the multiple loops. `mapValues` doesn't modify the original object passed in (the third parameter is modified, but it's initialized to an empty object by default, and normally isn't used). You should be able to replace the `remapObject` usage with it without any trouble. So I think this implementation doesn't have a reason to exist, as elegant as it appears.
2. Just letting you know about a Github quirk I just discovered thanks to your reply: Github doesn't place your reply in the correct place if the comment is an inline-code-comment, instead the reply goes to the PR, so to observers it looks weird.

sdegutis replied:
Good points, thanks.

sdegutis commented:
AFAIK there was a destructive version at the top of the file called mapValues which matches the name of it in lodash, but I needed an immutable one. But I could be wrong on both counts.

> On Oct 20, 2020, at 7:24 PM, Greg Slepak <notifications@github.com> wrote:
> 
> ﻿
> @taoeffect commented on this pull request.
> 
> In frontend/utils/giLodash.js:
> 
> > +export function remapObject (o, f) {
> +  return Object.fromEntries(Object.entries(o).map(([k, v]) => [k, f(v)]))
> +}
> Nicely done.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

taoeffect commented on frontend/utils/distribution/group-income-distribution.js:
```diff
@@ -1,40 +1,253 @@
+  const totalNeed = needs.reduce((a, b) => a + b.need, 0)
+  const totalPercent = Math.min(1, totalHave / totalNeed)
+
+  for (const have of haves) have.percent = have.have / totalHave
```
Does this need its own loop or can it be done inside of the other `for (const have of haves)` loop down below on line 66?

sdegutis replied:
We could recalculate it every time, sure. But I did this partly to help with debug-logging so that I could see the variables in the console.

taoeffect commented on frontend/utils/distribution/group-income-distribution.js:
```diff
@@ -1,40 +1,253 @@
+    else if (event.type === 'join') {
+      const newPayments = handleJoin({ user: event, needs })
+      payments = payments.concat(newPayments)
+    }
```
Minor: I have made it a habit to recognize if/else blocks like this, and write them in a way that is slightly different. Since if/else always creates a branch, and since branches slow things down (and can also make code slightly less readable in some situations), I use a map in situations like this:

```js
for (const event of events) {
  ({
    payment: event => handlePayment({ payments, needs, payment: event }),
    join: event => {
      const newPayments = handleJoin({ user: event, needs })
      payments = payments.concat(newPayments)
    }
  })[event.type](event)
}
```

It also has the added benefit of automatically handling unknown types by throwing an exception.

You don't need to adopt this method if you don't like it. I bring it up only as a neat trick I picked up from my C days of function tables for creating branchless code. It starts to pay off especially when there is a large number of events, and the body of code for each event is large.

sdegutis replied:
Thanks for letting me know. If you mean it slows things down in the JavaScript VM, looped function creation and execution is probably much slower than if-branching, although performance doesn't really matter here.

taoeffect replied:
Oh yeah, I don't know if the JS compiler would actually do that, but if it did then yeah it would make sense to move the definition of that object outside of the loop.

sdegutis committed:
[#763] Using mapValues instead of remapObject

sdegutis committed:
[#763] Changing Scenario 1 to represent added-income as "join" event.

taoeffect commented:
Closing in favor of #1009.
